Forms: Add REST API endpoint for exporting responses

lezama opened the pull request:
## Proposed changes
This PR replaces the legacy AJAX-based CSV export with a modern REST API endpoint for Jetpack Forms responses.

### Changes:
- Add new `/wp/v2/feedback/export` REST endpoint for CSV exports
- Remove legacy AJAX handler `wp_ajax_feedback_export` and the `download_feedback_as_csv()` method
- Update Jetpack Forms dashboard to use REST API with `@wordpress/api-fetch`
- Support all existing filters: selected items, search, status, date ranges, and source/parent

## Other information
<!-- This replaces the legacy AJAX implementation that was prone to nonce issues and required complex handling. The new REST API approach is more maintainable and follows WordPress standards. -->

## Testing instructions
<!-- Please explain how to test your changes. -->
1. Navigate to Jetpack → Forms in WP Admin
2. Test exporting with various filters:
   - Export all responses (no selection)
   - Select specific responses and export
   - Filter by source/form and export
   - Apply search filters and export
   - Filter by date range and export
3. Verify CSV downloads correctly with expected data
4. Confirm permission checks work (test with non-admin user)

## Product impact
<!-- Can you think of any product impact from this change? Please indicate how this change would impact users. -->
- No user-facing changes - export functionality remains the same
- Improved reliability and error handling for exports
- Better compatibility with modern WordPress environments

## Does this pull request change what data or activity we track or use?
<!-- If this pull request changes what data we track or use, please include a short explanation here, along with your DNSM ticket that covers the changes. -->
No

## Checklist
- [x] I have written a clear PR title and description
- [x] I have tested my changes
- [x] My code follows the WordPress coding standards
- [x] My code passes the linter

lezama committed:
Use rest endpoint instead of ajax

enejb commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -1070,4 +1107,158 @@ public function get_forms_config( WP_REST_Request $request ) { // phpcs:ignore V
+		$query_args = array(
+			'post_type'        => 'feedback',
+			'posts_per_page'   => -1,
+			'post_status'      => array( 'publish', 'draft' ),
```
I don't think we need to export "draft" feedback posts.

enejb commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -2756,84 +2754,6 @@ function ( $selected ) {
-	/**
-	 * Download exported data as CSV
-	 */
-	public function download_feedback_as_csv() {
```
We should deprecate this method first. In case it is used by something else.

lezama replied:
I made it possible to call it using the old signature, so it won’t break in case anyone is still using it.

enejb commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -1070,4 +1107,158 @@ public function get_forms_config( WP_REST_Request $request ) { // phpcs:ignore V
+			$post_data = array(
+				'ID'           => $post->ID,
+				'Date'         => get_the_date( 'Y-m-d H:i:s', $post ),
+				'Author Name'  => get_post_meta( $post->ID, '_feedback_author', true ),
```
I don't think such meta exists.  Or at least there shouldn't be a need for it to exist. Since the info is stored on the post content instead. 

Lets use the `get_export_feedback_data` method instead.

lezama replied:
yes

lezama committed:
not deprecating

manzoorwanijk commented:
> In this version, we return JSON for errors and CSV on success. While `admin_post_{$action}` is conventional, clients using `@wordpress/api-fetch` benefit from REST’s built-in auth handling, instead of adding boilerplate for an AJAX nonce. Is there a helper or way to use existing AJAX endpoints within `@wordpress` packages?

If you wish to laverage the REST authentication for api-fetch, then you can return a URL to admin-post action mentioned above. So, it can pass only the `$feedback_ids` and `$post_id` to the admin post which should make the admin post action simpler. You can also add a nonce to the URL.

Then update `Contact_Form_Plugin::download_feedback_as_csv()` to be a callback to `admin_post_feedback_export`, which verifies the nonce and receives the feedback IDs to export, which should be easy to sanitize. Then in API fetch, you can pass the content-type header or `parse: false` as you already did when you try to download the file.

manzoorwanijk commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -1070,4 +1107,106 @@ public function get_forms_config( WP_REST_Request $request ) { // phpcs:ignore V
+	 */
+	public function export_responses( $request ) {
+		$selected = $request->get_param( 'selected' );
+		$post_id  = $request->get_param( 'post' );
+		$search   = $request->get_param( 'search' );
+		$status   = $request->get_param( 'status' );
+		$before   = $request->get_param( 'before' );
+		$after    = $request->get_param( 'after' );
```
Don't we need to sanitize these values? The schema defines them as just strings, which can be anything, which may not be safe to pass directly to the query below.

So, may be `wp_unlash` and then `sanitize_text_field`?

lezama replied:
Added sanitize_callbacks 👍

lezama committed:
Refactor CSV export to use admin-post

lezama committed:
simpler

lezama committed:
add sanitize_calllbacks

lezama committed:
fix types

lezama committed:
simpler

lezama committed:
simpler

lezama committed:
Merge branch 'trunk' into forms/add-rest-export-endpoint

manzoorwanijk commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -1068,4 +1110,115 @@ public function get_forms_config( WP_REST_Request $request ) { // phpcs:ignore V
+			array(
+				'action'       => 'feedback_export',
+				'feedback_ids' => implode( ',', $feedback_ids ),
+				'post_id'      => $post_id,
```
We don’t seem to be using any of these params in admin post action. Do we need to pass them here?

lezama committed:
remove unused params

manzoorwanijk commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -2756,22 +2757,53 @@ function ( $selected ) {
-		$data      = $this->get_feedback_entries_from_post();
+	public function admin_post_feedback_export() {
+		$feedback_ids_str = sanitize_text_field( wp_unslash( $_GET['feedback_ids'] ?? '' ) );
+		$post_id          = sanitize_text_field( wp_unslash( $_GET['post_id'] ?? '' ) );
```
We don't pass the `post_id` when generating the URL from the REST endpoint. Do we?

manzoorwanijk commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -2756,22 +2757,53 @@ function ( $selected ) {
+		}
+
+		$feedback_ids = explode( ',', $feedback_ids_str );
+		$feedback_ids = array_map( 'intval', $feedback_ids );
```
May be clean up any empty values if present

```suggestion
		$feedback_ids = array_values( array_filter( array_map( 'intval', $feedback_ids ) ) );
```

Also, I suggest that we do that `'Invalid request parameters.'` check after this step, because `"0,0"` will pass that check above but won't pass if we check after this step.

manzoorwanijk commented on projects/packages/forms/src/dashboard/hooks/use-export-responses.ts:
```diff
@@ -75,25 +89,34 @@ export default function useExportResponses(): ExportHookReturn {
+			window.location.href = response.download_url;
+			return response;
+		}
+		throw new Error( 'Invalid response: missing download URL' );
```
Do we need to translate this?

enejb replied:
I would say not yet since it would only show up in the dev console for now.

manzoorwanijk commented on projects/packages/forms/src/contact-form/class-contact-form-endpoint.php:
```diff
@@ -1068,4 +1110,110 @@ public function get_forms_config( WP_REST_Request $request ) { // phpcs:ignore V
+			return new WP_Error( 'no_responses', __( 'No responses found', 'jetpack-forms' ), array( 'status' => 404 ) );
+		}
+
+		$feedback_ids = wp_list_pluck( $feedback_posts, 'ID' );
```
You can add [`'fields' => 'ids'`](https://developer.wordpress.org/reference/classes/wp_query/#return-fields-parameter) to `$query_args` above to get rid of this step and make it a bit faster.

manzoorwanijk reviewed (commented):
Thank you for implementing the suggestions. This looks much better now. I have some further suggestions for improvement, some of which you may ignore if you prefer.

simison committed:
Add export progress indicator to CSV export

Introduces an isExporting state to track export progress and prevent UI flickering. The export button is now disabled and shows a busy indicator while exporting, improving user feedback during CSV downloads. props @simison

simison commented:
>  It would be good to show some sort of in progress button while the URL is being returned. This will help the user not click the button multiple times.

[`Button`](https://wordpress.github.io/gutenberg/?path=/docs/components-button--docs) has `disabled` and `isBusy` states for this purpose, here's a PR adding `isExporting` flag on top of this PR, you can either merge into this or deploy separately, how you prefer: 
- https://github.com/Automattic/jetpack/pull/45302

enejb committed:
Update projects/packages/forms/src/contact-form/class-contact-form-plugin.php

Co-authored-by: Manzoor Wani <manzoorwani.jk@gmail.com>

enejb committed:
Use feedback ids instead

enejb committed:
Show error notice when there are no responses.

enejb committed:
Add tests

enejb committed:
Fix linter errors

enejb committed:
More tests

enejb committed:
Pass the $post_id

manzoorwanijk reviewed (dismissed):
Thank you for addressing the feedback. This looks good to me now.

<details><summary> 👉 I would still love to improve <code>admin_post_feedback_export</code> method in some next iteration like this.</summary>

```php
function admin_post_feedback_export() {
	// First check for the user permissions before we do anything.
	if ( ! current_user_can( 'export' ) ) {
		wp_die( esc_html__( 'You do not have permission to export form responses.', 'jetpack-forms' ), 403 );
	}

	// Next, let us sanitize the input parameters that we need immediately.
	$feedback_ids_str = sanitize_text_field( wp_unslash( $_GET['feedback_ids'] ?? '' ) );
	$nonce            = sanitize_text_field( wp_unslash( $_GET['nonce'] ?? '' ) );

	// Ensure we have the required parameters.
	if ( empty( $feedback_ids_str ) || empty( $nonce ) ) {
		wp_die( esc_html__( 'Invalid request parameters.', 'jetpack-forms' ), 400 );
	}

	// Verify the nonce.
	if ( ! wp_verify_nonce( $nonce, 'feedback_export_' . $feedback_ids_str ) ) {
		wp_die( esc_html__( 'Security check failed.', 'jetpack-forms' ), 403 );
	}

	// Now extract and sanitize feedback IDs.
	$feedback_ids = array_values(
		array_filter(
			array_map(
				'intval',
				explode( ',', $feedback_ids_str )
			)
		)
	);

	// Shortcircuit if we have no feedback IDs.
	if ( empty( $feedback_ids ) ) {
		wp_die( esc_html__( 'Invalid request parameters.', 'jetpack-forms' ), 400 );
	}

	// Let us get the export data.
	$export_data = $this->get_export_feedback_data( $feedback_ids );

	if ( empty( $export_data ) ) {
		wp_die( esc_html__( 'No responses found to export.', 'jetpack-forms' ), 404 );
	}

	// Now let us get the post ID as we need it now.
	$post_id = sanitize_text_field( wp_unslash( $_GET['post_id'] ?? '' ) );

	$this->download_feedback_as_csv( $export_data, $post_id );
}
```
</details>

enejb committed:
Keep the previous method as it was

edanzer commented on projects/packages/forms/src/dashboard/hooks/use-export-responses.ts:
```diff
@@ -75,25 +98,47 @@ export default function useExportResponses(): ExportHookReturn {
+			} );
+			setIsExporting( false );
+		}
+	}, [ currentQuery, selected, closeModal, createErrorNotice ] );
```
We've updated the onExport method returned from this hook. It doesn’t accept args and returns an  object from the endpoint rather than fetch Response. 

This onExport method is passed to both the CSV and Google export components, and the CSV file has been updated to use the new onExport as expected. 

But the Google Drive card is still using this same onExport function the old way [here](https://github.com/Automattic/jetpack/blob/1f5164adcd57102229bef99d5afdbbd9469b420c/projects/packages/forms/src/dashboard/inbox/export-responses/google-drive.tsx#L43). 

```
onExport( 'grunion_export_to_gdrive', 'feedback_export_nonce_gdrive' )
			.then( ( response: Response ) => response.json() )
			.then( ( { data } ) => {
				window.open( data.sheet_link, '_blank' );
			} )
			.finally( () => {
				setIsExporting( false );
			} );
	}, [ tracks, onExport, isExporting ] );
```

I'd think we either need to update Google drive to use the new onExport method correctly, or else if there's a reason we don't want to tackle that here, maybe have the useExportResponses() hook still return the old method (for Google) along with the new one (for CSV), and then handle updating Google in a follow up.

edanzer replied:
To confirm if you try to Export to Google with this PR, you do get an error: 

```
Uncaught (in promise) TypeError: response.json is not a function
    at eval (google-drive.tsx:76:100)
```

enejb replied:
Good catch. @edanzer. It used to work but I think I broke it :( 

Will create a fix for google drive.

lezama replied:
thanks!
